File: Skydown App/Services/AI/AgentPendingQueueStore.swift
import Foundation

struct AgentPendingQueueTurn: Codable, Equatable {
    let role: String
    let text: String
}

struct AgentPendingQueueEntry: Codable, Equatable {
    let userKey: String
    let sessionID: String
    let prompt: String
    let history: [AgentPendingQueueTurn]
    let mode: String
    let aiLevel: String
    let executeAutomation: Bool
    let automationScope: String
    let assistantMessageID: String
    let createdAt: Date
    let attachments: [AgentOutboundAttachment]

    init(
        userKey: String,
        sessionID: String,
        prompt: String,
        history: [AgentPendingQueueTurn],
        mode: String,
        aiLevel: String,
        executeAutomation: Bool,
        automationScope: String = "owner",
        assistantMessageID: String,
        createdAt: Date,
        attachments: [AgentOutboundAttachment] = []
    ) {
        self.userKey = userKey
        self.sessionID = sessionID
        self.prompt = prompt
        self.history = history
        self.mode = mode
        self.aiLevel = aiLevel
        self.executeAutomation = executeAutomation
        self.automationScope = automationScope
        self.assistantMessageID = assistantMessageID
        self.createdAt = createdAt
        self.attachments = attachments
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        userKey = try container.decode(String.self, forKey: .userKey)
        sessionID = try container.decodeIfPresent(String.self, forKey: .sessionID) ?? ""
        prompt = try container.decode(String.self, forKey: .prompt)
        history = try container.decode([AgentPendingQueueTurn].self, forKey: .history)
        mode = try container.decode(String.self, forKey: .mode)
        aiLevel = try container.decodeIfPresent(String.self, forKey: .aiLevel) ?? AIExperienceLevel.standard.rawValue
        executeAutomation = try container.decode(Bool.self, forKey: .executeAutomation)
        automationScope = try container.decodeIfPresent(String.self, forKey: .automationScope) ?? "owner"
        assistantMessageID = try container.decode(String.self, forKey: .assistantMessageID)
        createdAt = try container.decode(Date.self, forKey: .createdAt)
        attachments = try container.decodeIfPresent([AgentOutboundAttachment].self, forKey: .attachments) ?? []
    }

    private enum CodingKeys: String, CodingKey {
        case userKey
        case sessionID
        case prompt
        case history
        case mode
        case aiLevel
        case executeAutomation
        case automationScope
        case assistantMessageID
        case createdAt
        case attachments
    }
}

@MainActor
final class AgentPendingQueueStore {
    static let shared = AgentPendingQueueStore()

    private let key = "agent_pending_queue_entries"
    private let maximumEntries = 80
    private let maximumAgeDays = 14
    private let defaults: UserDefaults

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        pruneExpiredEntries()
    }

    func entries(for userKey: String?, sessionID: String?) -> [AgentPendingQueueEntry] {
        let normalizedUserKey = normalizeUserKey(userKey)
        let normalizedSessionID = normalizeSessionID(sessionID)
        return readEntries()
            .filter {
                $0.userKey == normalizedUserKey &&
                normalizeSessionID($0.sessionID) == normalizedSessionID
            }
            .sorted { $0.createdAt < $1.createdAt }
    }

    func saveEntries(_ entries: [AgentPendingQueueEntry], for userKey: String?, sessionID: String?) {
        let normalizedUserKey = normalizeUserKey(userKey)
        let normalizedSessionID = normalizeSessionID(sessionID)
        let others = readEntries().filter {
            !($0.userKey == normalizedUserKey && normalizeSessionID($0.sessionID) == normalizedSessionID)
        }
        let sanitized = entries
            .filter { !$0.prompt.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .map {
                AgentPendingQueueEntry(
                    userKey: normalizedUserKey,
                    sessionID: normalizedSessionID,
                    prompt: $0.prompt,
                    history: $0.history,
                    mode: $0.mode,
                    aiLevel: $0.aiLevel,
                    executeAutomation: $0.executeAutomation,
                    automationScope: $0.automationScope,
                    assistantMessageID: $0.assistantMessageID,
                    createdAt: $0.createdAt,
                    attachments: $0.attachments
                )
            }
            .sorted { $0.createdAt < $1.createdAt }

        var merged = others + sanitized
        merged.sort { $0.createdAt > $1.createdAt }
        merged = Array(merged.prefix(maximumEntries))
        writeEntries(merged)
        pruneExpiredEntries()
    }

    func clearEntries(for userKey: String?, sessionID: String?) {
        saveEntries([], for: userKey, sessionID: sessionID)
    }

    func migrateLegacyEntries(for userKey: String?, to sessionID: String?) {
        let normalizedUserKey = normalizeUserKey(userKey)
        let normalizedSessionID = normalizeSessionID(sessionID)
        guard !normalizedSessionID.isEmpty else { return }

        let migratedEntries = readEntries().map { entry in
            guard entry.userKey == normalizedUserKey,
                  normalizeSessionID(entry.sessionID).isEmpty else {
                return entry
            }

            return AgentPendingQueueEntry(
                userKey: normalizedUserKey,
                sessionID: normalizedSessionID,
                prompt: entry.prompt,
                history: entry.history,
                mode: entry.mode,
                aiLevel: entry.aiLevel,
                executeAutomation: entry.executeAutomation,
                automationScope: entry.automationScope,
                assistantMessageID: entry.assistantMessageID,
                createdAt: entry.createdAt,
                attachments: entry.attachments
            )
        }

        if migratedEntries != readEntries() {
            writeEntries(migratedEntries)
        }
    }

    private func pruneExpiredEntries() {
        let cutoff = Calendar.current.date(byAdding: .day, value: -maximumAgeDays, to: .now) ?? .distantPast
        let existingEntries = readEntries()
        let filteredEntries = existingEntries.filter { $0.createdAt >= cutoff }
        if filteredEntries != existingEntries {
            writeEntries(filteredEntries)
        }
    }

    private func readEntries() -> [AgentPendingQueueEntry] {
        guard let data = defaults.data(forKey: key) else { return [] }
        return (try? JSONDecoder().decode([AgentPendingQueueEntry].self, from: data)) ?? []
    }

    private func writeEntries(_ entries: [AgentPendingQueueEntry]) {
        guard let data = try? JSONEncoder().encode(entries) else { return }
        defaults.set(data, forKey: key)
    }

    private func normalizeUserKey(_ userKey: String?) -> String {
        let trimmed = userKey?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        return trimmed.isEmpty ? "guest" : trimmed.lowercased()
    }

    private func normalizeSessionID(_ sessionID: String?) -> String {
        sessionID?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased() ?? ""
    }
}

File: androidApp/src/main/java/com/nash/skyos/data/AgentPendingQueueStore.kt
package com.nash.skyos.data

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

data class AgentPendingQueueEntry(
    val userKey: String,
    val sessionId: String,
    val prompt: String,
    val history: List<AgentHistoryTurn>,
    val mode: String,
    val aiLevel: String = "standard",
    val executeAutomation: Boolean,
    val automationScope: String = "owner",
    val assistantMessageId: String,
    val createdAtEpochMillis: Long = System.currentTimeMillis(),
    val attachments: List<AgentOutboundAttachment> = emptyList(),
    val idempotencyKey: String = "",
)

object AgentPendingQueueStore {
    private const val preferencesName = "agent_pending_queue"
    private const val entriesKey = "entries"
    private const val maximumEntries = 80
    private const val maximumAgeMillis = 14L * 24L * 60L * 60L * 1000L

    private lateinit var sharedPreferences: SharedPreferences

    fun initialize(context: Context) {
        if (::sharedPreferences.isInitialized) {
            return
        }
        sharedPreferences = context.getSharedPreferences(preferencesName, Context.MODE_PRIVATE)
        pruneExpiredEntries()
    }

    fun entriesFor(userKey: String?, sessionId: String?): List<AgentPendingQueueEntry> {
        val normalized = normalizeUserKey(userKey)
        val normalizedSessionId = normalizeSessionId(sessionId)
        return readEntries()
            .filter { it.userKey == normalized && it.sessionId == normalizedSessionId }
            .sortedBy { it.createdAtEpochMillis }
    }

    fun saveEntriesForSession(userKey: String?, sessionId: String?, entries: List<AgentPendingQueueEntry>) {
        val normalized = normalizeUserKey(userKey)
        val normalizedSessionId = normalizeSessionId(sessionId)
        val others = readEntries().filterNot { it.userKey == normalized && it.sessionId == normalizedSessionId }
        val sanitized = entries
            .filter { it.prompt.isNotBlank() && it.assistantMessageId.isNotBlank() }
            .sortedBy { it.createdAtEpochMillis }
            .map { entry ->
                entry.copy(
                    userKey = normalized,
                    sessionId = normalizedSessionId,
                    assistantMessageId = entry.assistantMessageId.ifBlank { UUID.randomUUID().toString() },
                )
            }

        val merged = (others + sanitized)
            .sortedByDescending { it.createdAtEpochMillis }
            .take(maximumEntries)
        writeEntries(merged)
        pruneExpiredEntries()
    }

    fun clearEntriesForSession(userKey: String?, sessionId: String?) {
        saveEntriesForSession(userKey, sessionId, emptyList())
    }

    private fun pruneExpiredEntries() {
        val cutoff = System.currentTimeMillis() - maximumAgeMillis
        val filtered = readEntries().filter { it.createdAtEpochMillis >= cutoff }
        writeEntries(filtered)
    }

    private fun readEntries(): List<AgentPendingQueueEntry> {
        check(::sharedPreferences.isInitialized) {
            "AgentPendingQueueStore.initialize(context) must be called before reading."
        }
        val raw = sharedPreferences.getString(entriesKey, null).orEmpty()
        if (raw.isBlank()) {
            return emptyList()
        }

        return runCatching {
            val array = JSONArray(raw)
            buildList {
                for (index in 0 until array.length()) {
                    val item = array.optJSONObject(index) ?: continue
                    val prompt = item.optString("prompt").trim()
                    if (prompt.isBlank()) continue

                    val assistantMessageId = item.optString("assistantMessageId")
                        .trim()
                        .ifBlank { UUID.randomUUID().toString() }
                    val history = parseHistory(item.optJSONArray("history"))

                    add(
                        AgentPendingQueueEntry(
                            userKey = normalizeUserKey(item.optString("userKey")),
                            sessionId = normalizeSessionId(item.optString("sessionId")),
                            prompt = prompt,
                            history = history,
                            mode = item.optString("mode").trim().ifBlank { "release" },
                            aiLevel = item.optString("aiLevel").trim().ifBlank { "standard" },
                            executeAutomation = item.optBoolean("executeAutomation", false),
                            automationScope = item.optString("automationScope").trim().ifBlank { "owner" },
                            assistantMessageId = assistantMessageId,
                            createdAtEpochMillis = item.optLong("createdAtEpochMillis", System.currentTimeMillis()),
                            attachments = parseAttachments(item.optJSONArray("attachments")),
                            idempotencyKey = item.optString("idempotencyKey").trim(),
                        ),
                    )
                }
            }
        }.getOrDefault(emptyList())
    }

    private fun parseAttachments(array: JSONArray?): List<AgentOutboundAttachment> {
        if (array == null) return emptyList()
        return buildList {
            for (index in 0 until array.length()) {
                val o = array.optJSONObject(index) ?: continue
                val name = o.optString("name").trim()
                if (name.isBlank()) continue
                val b64 = o.optString("inlineBase64").trim()
                if (b64.isBlank()) continue
                add(
                    AgentOutboundAttachment(
                        id = o.optString("id").trim().ifBlank { "att_${size + 1}" },
                        name = name,
                        kind = o.optString("kind").trim().ifBlank { "file" },
                        mimeType = o.optString("mimeType").trim().ifBlank { "application/octet-stream" },
                        source = o.optString("source").trim().ifBlank { "inline" },
                        inlineBase64 = b64,
                    ),
                )
            }
        }
    }

    private fun parseHistory(array: JSONArray?): List<AgentHistoryTurn> {
        if (array == null) return emptyList()
        return buildList {
            for (index in 0 until array.length()) {
                val item = array.optJSONObject(index) ?: continue
                val role = item.optString("role").trim()
                val text = item.optString("text").trim()
                if (role.isBlank() || text.isBlank()) {
                    continue
                }
                add(AgentHistoryTurn(role = role, text = text))
            }
        }
    }

    private fun writeEntries(entries: List<AgentPendingQueueEntry>) {
        check(::sharedPreferences.isInitialized) {
            "AgentPendingQueueStore.initialize(context) must be called before writing."
        }
        val array = JSONArray()
        entries.forEach { entry ->
            array.put(
                JSONObject()
                    .put("userKey", normalizeUserKey(entry.userKey))
                    .put("sessionId", normalizeSessionId(entry.sessionId))
                    .put("prompt", entry.prompt)
                    .put("mode", entry.mode)
                    .put("aiLevel", entry.aiLevel.ifBlank { "standard" })
                    .put("executeAutomation", entry.executeAutomation)
                    .put("automationScope", entry.automationScope.ifBlank { "owner" })
                    .put("assistantMessageId", entry.assistantMessageId)
                    .put("createdAtEpochMillis", entry.createdAtEpochMillis)
                    .put("idempotencyKey", entry.idempotencyKey)
                    .put(
                        "history",
                        JSONArray().apply {
                            entry.history.forEach { turn ->
                                put(
                                    JSONObject()
                                        .put("role", turn.role)
                                        .put("text", turn.text),
                                )
                            }
                        },
                    )
                    .put(
                        "attachments",
                        JSONArray().apply {
                            entry.attachments.forEach { att ->
                                put(
                                    JSONObject()
                                        .put("id", att.id)
                                        .put("name", att.name)
                                        .put("kind", att.kind)
                                        .put("mimeType", att.mimeType)
                                        .put("source", att.source)
                                        .put("inlineBase64", att.inlineBase64),
                                )
                            }
                        },
                    ),
            )
        }
        sharedPreferences.edit().putString(entriesKey, array.toString()).apply()
    }

    private fun normalizeUserKey(userKey: String?): String {
        val trimmed = userKey?.trim().orEmpty()
        return trimmed.ifBlank { "guest" }.lowercase()
    }

    private fun normalizeSessionId(sessionId: String?): String {
        return sessionId?.trim().orEmpty()
    }
}
